### ISSUE TITLE
plugin architecture: deploy, system, and site level plugin design

### ISSUE THREAD

**Issue by btopro** [haxtheweb/issues#1325]
site level data based apps need a way to store their data WITH the site itself. This idea has come up before but we have a lot of the wiring to actually do it now.

- site level would be like a commenting app which writes into a folder like `/sites/SITENAME/appData/inline-discussion/data.json`
- endpoint then needs built into haxcms (for now) which then does the writing
- similiar process for user data but it's things like "personal notes" which can just be activated at the site level (some how) or built directly into haxcms

![image](https://user-images.githubusercontent.com/329735/233073241-c0947d82-693d-46cb-81a2-54a3e9a88ce4.png)

**Comment by btopro** [comment]
#2769

**Comment by btopro** [comment]
A plugin would be autoloaded from a directory and it would need to have the following characteristics:
- folder under haxiam/config/plugins (deployment level), config/plugins (system level) or {site}/config/plugins (site specific)
- yaml for info file
- openapi spec for endpoints if extending endpoints of the system
- area for js based code for frontend / blocks
- nodejs and php based integration
- endpoint `api/v1/plugins` registry which can be used for highlighting which plugins are enabled / disabled

Candidates for converting to a plugin architecture in order to build out future functionality:
- youtube / appstore brokers currently hard coded
- haxiam user access endpoints (sharing resources without hard coding the plugin as currently)
- xAPI / user progress tracking beacon and storage
- commenting engines for sites
- other future integrations meant to extend the data model / capabilities
- rewriting of the core elements to be plugin derived (site, system, session being the big three currently)

This can start to work its way into the platform capabilities as far as extending these capabilities for what's enables / disabled. Then a user could enable "youtube" (if disabled) and supply their API key for it to work (settings would have to be able to be served via open API spec). These are very complex documents with conformance requirements -- which is why the world of AI make it possible to require them 🎆

**Comment by btopro** [comment]
## Implementation plan

> Drafted via the HAX Issue Research workflow — grounded in local codebase research, updated with resolved open questions from the issue author. This is an ideation-phase architectural plan; implementation will follow once the approach is ratified.

**Issue:** https://github.com/haxtheweb/issues/issues/1325
**Repo / surface:** `haxcms-nodejs` (primary) + `haxcms-php` (parity) + `HAXiam` (deployment-level config) + `webcomponents` (frontend block registration)

### Problem statement
HAXcms needs a plugin architecture that allows extensions (app-store brokers, commenting engines, xAPI tracking, HAXiam user access, etc.) to be autoloaded from directories at three scopes (site, system, core), declare their own endpoints via OpenAPI fragments, supply frontend blocks, provide settings forms, enforce Drupal-style dependencies, support lifecycle hooks, and be enabled/disabled per site via the platform capabilities system. Currently these extensions are hard-coded (e.g., `PROVIDER_DEFINITIONS` in `integrations.js`, `loadBaseAppStore()` in `HAXAppStoreService.js`).

### Current state / context
- **Skeleton system is the closest existing analog**: `haxcms-nodejs/src/lib/skeletonSettings.js:58-73` (`getSkeletonDirectories()`) implements a 3-tier cascade: `user` (`configDirectory/user/skeletons`), `config` (`configDirectory/skeletons`), `core` (`coreConfigPath/skeletons`). `discoverSkeletons()` (line 184) walks these dirs, loads JSON, deduplicates by machine name (first scope wins). `enabledSkeletons.json` in settings tracks enable/disable state. **This is the exact pattern plugins should follow**, extended with executable code and OpenAPI fragments.
- **Config discovery cascade**: `haxcms-nodejs/src/lib/discoverConfigPath.js` walks from cwd to root looking for `_config`/`.haxcmsconfig` dirs with `.isHAXcmsConfig` marker. Home dir fallback. Creates `settings/`, `user/skeletons/`, `skeletons/` dirs on first run (line 38-53).
- **Deployment profiles**: `haxcms-nodejs/src/lib/HAXCMS.js:3096-3105` — `getDeploymentProfile()` returns `single-site`, `self-hosted-multi-site`, or `haxiam-managed`. These control which plugin scopes are active.
- **Platform features (enable/disable)**: `haxcms-nodejs/src/lib/platformFeatures.js` — `platformAllows(site, capability)` reads from `site.manifest.metadata.platform.features`. Issue #2478 (closed) established the `manifest.metadata.platform` pattern. Plugins register capabilities that flow through this same mechanism.
- **API keys (settings pattern)**: `haxcms-nodejs/src/lib/apiKeys.js` — `SUPPORTED_API_KEY_PROVIDERS` is hard-coded. Reads/writes `settings/apiKeys.json`. This pattern generalizes to plugin-supplied settings.
- **App-store brokers (prime conversion candidate)**: `haxcms-nodejs/src/siteRoutes/v1/integrations.js:19-90` has `PROVIDER_DEFINITIONS` (hard-coded). `haxcms-nodejs/src/lib/HAXAppStoreService.js:10-711` has `loadBaseAppStore()` with embedded JSON strings. **All of this should become plugin-supplied.**
- **Route registration (static)**: `haxcms-nodejs/src/lib/SiteRoutesMap.js:88-108` (`V1_ENTITY_ROUTE_DEFINITIONS`) + `SystemRoutesMap.js` — routes statically registered at startup via `addRouteHandler()`. Plugins dynamically inject routes.
- **OpenAPI spec drives auth**: `haxcms-nodejs/src/app.js:1541-1606` — `readSiteApiAuthPoliciesFromOpenApiSpec()` parses `site-spec.yaml` to determine auth policies per route. Routes not in the spec fail-closed to `authenticated`.
- **Frontend element loading**: `haxcms-nodejs/src/coreConfig/dynamicElementLoader.json` maps tag names to import paths. Plugins contribute entries.
- **PHP parity**: `haxcms-php/system/api.php` uses `SystemApiRouter::dispatch()`. The PHP `_config` directory mirrors the NodeJS config directory.
- **Issue #213 vision**: Plugin distribution as a decentralized "supplier" and "package manager" ecosystem — like apt-get/yum for HAX. Different suppliers for different contexts (EDU, cooking, etc.). CDN alignment with supplier-specific configurations. Start on-prem, work up to registry.

### Proposed approach

**1. Plugin directory cascade (site → system → core)**

Plugins autoloaded from three directories, first-match-wins by machine name:
- **Site scope**: `{site.siteDirectory}/config/plugins` — per-site plugin overrides
- **System scope**: `{configDirectory}/plugins` — system-wide plugins (self-hosted or HAXiam)
- **Core scope**: `{coreConfigPath}/plugins` — built-in plugins shipped with HAXcms

A small core of always-loaded **required** plugins ensures the system works. Other plugins are optional and can be enabled/disabled. This enables refactoring parts of the codebase to be loaded as plugins with dependency management, while keeping essential functionality guaranteed.

New `haxcms-nodejs/src/lib/pluginLoader.js` (mirroring `skeletonSettings.js`) implements `getPluginDirectories()`, `discoverPlugins()`, `readEnabledPluginMap()`, `writeEnabledPluginMap()`.

**2. Plugin structure (per folder)**

```
plugins/
  comment-engine/
    plugin.yml              # info, capabilities, dependencies, required flag
    openapi.yaml            # OpenAPI 3.0 fragment: this plugin's own paths/schemas
    settings.schema.json    # HAXSchema form definition for plugin settings
    node/
      index.js              # exports: routes, hooks, capabilities
      handlers/             # route handler modules
    php/
      CommentEngine.php     # PHP parity handler
    frontend/
      blocks/               # web component JS files
      elements.json         # tag-name -> import-path mappings
```

`plugin.yml` fields:
- `machineName` — unique identifier (sanitized, kebab-case)
- `title`, `description`, `version`, `author`
- `scope` — site/system/core
- `required` — boolean; if true, always loaded and cannot be disabled
- `capabilities` — list of platform capability keys (namespaced, e.g., `comment-engine.hasComments`)
- `dependencies` — list of other plugin machine names this plugin requires
- `settingsSchema` — path to settings.schema.json
- `openApiSpec` — path to openapi.yaml
- `languages` — list of supported backend languages (e.g., `['node', 'php']`); a plugin can be Node-only, PHP-only, or both

**3. Generic plugin route dispatch (key architectural decision)**

Plugin routes use a **generic dispatch pattern** rather than merging all plugin OpenAPI specs into one master spec:

- The main `site-spec.yaml` declares a generic path: `/x/api/v1/plugin/{pluginMachineName}/{callback}` with a base security policy.
- When a request matches this pattern, the backend extracts `pluginMachineName` and the sub-path, looks up the plugin's route map (registered by its `node/index.js`), and dispatches to the plugin's handler.
- The plugin's own `openapi.yaml` is **consulted internally** by the backend for request/response schema validation and for per-endpoint auth policy. It is **not merged** into the main spec.
- Each plugin's OpenAPI spec is served at a discovery endpoint: `GET /x/api/v1/plugin/{pluginMachineName}/openapi.json` (and `.yaml`) so clients and LLMs can understand the plugin's specific API surface.

This keeps the main spec small and generic while allowing each plugin to define rich, self-contained API contracts. The auth middleware (`validateSiteApiRouteAccess`) is extended so the plugin loader populates the `siteApiAuthPoliciesByMethodAndRoute` map with per-plugin-route entries extracted from each plugin's OpenAPI security declarations. This way the existing fail-closed enforcement works unchanged.

**4. Plugin registry API (new system-level endpoints)**

- `GET /system/api/v1/plugins` — list all discovered plugins with scope, enabled/disabled status, required flag, dependencies, settings summary
- `GET /system/api/v1/plugins/:pluginName` — plugin detail with full info, settings form schema, OpenAPI spec link
- `PATCH /system/api/v1/plugins/:pluginName` — enable/disable plugin (refuses to disable required plugins; refuses to enable if dependencies are missing — Drupal-style enforcement)
- `GET /system/api/v1/plugins/:pluginName/settings` — get plugin settings values + HAXSchema form
- `PATCH /system/api/v1/plugins/:pluginName/settings` — save plugin settings (validates against schema)
- `GET /system/api/v1/plugins/:pluginName/openapi.json` — plugin's OpenAPI spec (JSON)
- `GET /system/api/v1/plugins/:pluginName/openapi.yaml` — plugin's OpenAPI spec (YAML)

Registered in `SystemRoutesMap.js` as admin routes (require system dashboard access).

**5. Dependency enforcement (Drupal-style)**

- A plugin cannot be **enabled** if its declared dependencies are not enabled. The registry endpoint returns a 409 with a message listing missing dependencies.
- A plugin cannot be **disabled** if other enabled plugins depend on it. The registry endpoint returns a 409 with a message listing dependent plugins that must be disabled first.
- Required plugins (`required: true` in `plugin.yml`) cannot be disabled at all.
- Dependency resolution is recursive: enabling A which depends on B which depends on C requires both B and C to be enabled. The enable endpoint can optionally auto-enable dependencies (with a `?autoEnableDependencies=true` query param) or require manual enabling in order.

**6. Lifecycle hooks (Drupal 7 hook-style)**

A plugin's `node/index.js` exports a `hooks` object with named hook functions. The `pluginLoader.js` maintains a hook registry and invokes all enabled plugins' hooks for each event. Proposed hook points for MVP:

- `onSiteCreated(site, context)` — after a new site is created
- `onSiteDeleted(siteName, context)` — before/after a site is deleted
- `onSiteCloned(sourceSite, newSite, context)` — after a site is cloned
- `onPageCreated(site, item, context)` — after a new page is created
- `onPageSaved(site, item, context)` — after a page's content/metadata is saved
- `onPageDeleted(site, itemId, context)` — after a page is deleted
- `onManifestSaved(site, context)` — after site.json manifest is saved
- `onOutlineSaved(site, items, context)` — after the outline structure is saved
- `onFileUploaded(site, file, context)` — after a file is uploaded
- `onFileDeleted(site, filePath, context)` — after a file is deleted
- `onUserLogin(userName, context)` — after a user logs in
- `onUserLogout(userName, context)` — after a user logs out
- `onPluginEnabled(pluginMachineName, context)` — after a plugin is enabled
- `onPluginDisabled(pluginMachineName, context)` — after a plugin is disabled

Hooks are fire-and-forget (async, errors logged but don't block the triggering operation). This matches Drupal 7's hook architecture where modules register callbacks invoked at specific points in the system lifecycle.

**7. Platform capabilities integration**

Plugins declare `capabilities` in `plugin.yml`. When enabled for a site, these flow into `site.manifest.metadata.platform` (or a new `site.manifest.metadata.plugins` namespace). The existing `platformAllows(site, capability)` function is extended to check plugin-provided capabilities. Example: `comment-engine` provides `hasComments`; frontend checks `store.platformAllows('hasComments')` before rendering.

**8. Settings management (plugin-supplied forms)**

Plugin `settings.schema.json` is a HAXSchema form definition. Plugin settings endpoints serve this as a form (same pattern as `siteSettingsForm`). Settings stored in `settings/pluginSettings/{pluginName}.json`. The `apiKeys.js` pattern is generalized — instead of a hard-coded `SUPPORTED_API_KEY_PROVIDERS`, a provider plugin's settings schema declares its `apiKey` field.

**9. OpenAPI fragment handling (per plugin, not merged)**

Plugin OpenAPI specs are **not merged** into the main spec. Instead:
- The main `site-spec.yaml` has the generic `/x/api/v1/plugin/{pluginMachineName}/{callback}` path with base security.
- Each plugin's `openapi.yaml` is self-contained with its own paths, schemas, and security schemes.
- At startup, the plugin loader reads each plugin's OpenAPI security declarations and populates the auth policy map so `validateSiteApiRouteAccess` enforces the correct policy per plugin route.
- Plugin OpenAPI specs are served at `/x/api/v1/plugin/{pluginMachineName}/openapi.json` and `.yaml` for discovery.
- The existing `/x/api/openapi` endpoint continues to serve the main spec (which now includes the generic plugin path pattern and a `plugins` directory listing linking to each plugin's spec).

**10. Plugin execution (main process, trusted)**

Plugin NodeJS modules run in the main process (no VM sandboxing). Plugins are trusted as part of the community/platform ecosystem — installed from a future registry or manually downloaded by developers. Core plugins shipped with the platform leverage the same architecture. This matches how `require()` already works for core modules. Provenance verification (checksums, signed manifests) is a future enhancement.

**11. Plugin distribution (on-prem first, registry later per #213)**

Per the issue author's direction and issue #213, plugin distribution starts with on-prem loading (manual copy / git clone / npm install into the cascade directories) and works up to a registry/supplier ecosystem. The directory structure supports all distribution methods. Issue #213's vision of decentralized "suppliers" and "package managers" (like apt-get/yum) is the long-term goal — different suppliers for different contexts (EDU, cooking, etc.) with CDN alignment. The MVP focuses on the loading architecture; the registry/supplier UI is future work.

**12. First conversion candidates (proof of concept)**

Per the issue comment, these hard-coded features become plugins:
- **YouTube/appstore brokers** -> `app-store-brokers` plugin (settings.schema.json for API keys, OpenAPI fragment for search endpoints, Node handler reusing existing `integrations.js` logic)
- **Commenting engine** -> `comment-engine` plugin (bridge to issue #2474)
- **xAPI/user progress tracking** -> `xapi-tracker` plugin
- **HAXiam user access/sharing** -> `haxiam-access` plugin (PHP-only initially, Node later — demonstrates language-specific handler support)

Core routes (site, system, session) remain hard-coded initially. A small core of required plugins (including the appDataStorage plugin from #2474) is always loaded.

### Files likely affected (new)
- `haxcms-nodejs/src/lib/pluginLoader.js` — **NEW** — plugin discovery, loading, cascade logic, dependency enforcement, hook registry
- `haxcms-nodejs/src/lib/pluginSettings.js` — **NEW** — plugin settings storage (generalizes `apiKeys.js`)
- `haxcms-nodejs/src/lib/pluginRouteDispatcher.js` — **NEW** — generic `/x/api/v1/plugin/:pluginMachineName/*` route handler that dispatches to plugin handlers
- `haxcms-nodejs/src/systemRoutes/v1/plugins.js` — **NEW** — plugin registry CRUD + settings + OpenAPI endpoints
- `haxcms-nodejs/src/coreConfig/plugins/` — **NEW** — directory for built-in core plugins
- `haxcms-nodejs/src/coreConfig/plugins/app-store-brokers/` — **NEW** — first converted plugin

### Files likely affected (modified)
- `haxcms-nodejs/src/app.js` — call plugin discovery + route registration + hook registration at startup (~line 1181 area); register generic plugin dispatch route
- `haxcms-nodejs/src/lib/SiteRoutesMap.js` — support late route registration from plugins; add generic plugin dispatch path
- `haxcms-nodejs/src/lib/SystemRoutesMap.js` — add plugin registry routes
- `haxcms-nodejs/src/lib/HAXCMS.js` — ensure `config/plugins` and `settings/pluginSettings` dirs exist (mirrors skeleton dir creation at line 2816-2828); add hook invocation points at key lifecycle events (site create/delete, page save/delete, user login/logout, etc.)
- `haxcms-nodejs/src/lib/platformFeatures.js` — extend `platformAllows()` to check plugin capabilities
- `haxcms-nodejs/src/lib/apiKeys.js` — generalize to read supported providers from discovered plugins
- `haxcms-nodejs/src/siteRoutes/v1/integrations.js` — refactor to load provider definitions from `app-store-brokers` plugin
- `haxcms-nodejs/src/lib/HAXAppStoreService.js` — refactor `loadBaseAppStore()` to load from plugin data
- `haxcms-nodejs/src/openapi/site-spec.yaml` — add generic `/x/api/v1/plugin/{pluginMachineName}/{callback}` path with base security

### Considerations / risks
- **DDD design system**: N/A for backend. Frontend blocks shipped by plugins should use DDD tokens.
- **Accessibility/WCAG**: Plugin-supplied frontend blocks must follow the same accessibility audits as core elements.
- **HAXSchema/HAX capability**: Plugin settings forms use HAXSchema. Plugin-supplied blocks must have `haxProperties` to be authorable in HAX.
- **New elements via CLI**: Plugin frontend blocks that are new web components must be created via `hax webcomponent` CLI, never by hand.
- **Backend parity (PHP)**: The issue requires "nodejs and php based integration." Each plugin has a `php/` directory with a PHP handler. Plugins can be Node-only, PHP-only, or both — the shared spec (plugin.yml + openapi.yaml) provides a unified contract. Language-specific handler directories allow developing one side then adding the other.
- **NO top-level monorepo build** (standing rule).
- **NO optional chaining (`?.`)** — the Polymer parser breaks on it (standing rule).
- **`globalThis` over `window`** for global-scope JS (standing rule).
- **Security — plugin code trusted in main process**: Plugins run in the main Node process with full access. For now, plugins are trusted as community/platform code — same trust model as npm packages or composer packages. Provenance verification (checksums, signed manifests) is a future enhancement.
- **Route name collision**: Since routing is namespaced by `pluginMachineName` (`/x/api/v1/plugin/{pluginMachineName}/{callback}`), collisions between plugins are impossible. Within a single plugin, the plugin's route map must not have duplicate keys — the loader validates this.
- **Capability namespace collision**: Plugin capability keys are namespaced by plugin machine name (e.g., `comment-engine.hasComments`) to avoid collisions with core capabilities (`hasDelete`, `hasAddPage`).
- **Reserved `x/` routes**: Plugin routes use `/x/api/v1/plugin/{pluginMachineName}/...` which follows the `/x/api/*` pattern and adds a `plugin/` namespace — no conflict with reserved `x/` internal paths like `x/search` or `x/tags`.
- **site.json `metadata.site.name` alignment**: Site-scoped plugins live under `{site.siteDirectory}/config/plugins` which is already aligned with the site name.
- **Performance**: Plugin discovery at startup is cached. The discovered plugin list, parsed OpenAPI specs, and hook registry are cached in memory. Invalidation on admin enable/disable or file watcher event.
- **Migration path**: Existing hard-coded brokers continue to work during transition. The `app-store-brokers` plugin is created first, then `integrations.js` and `HAXAppStoreService.js` are refactored to load from it. No breaking changes to existing API consumers.
- **Hook error isolation**: A failing hook in one plugin must not break the triggering operation or other plugins' hooks. Hook invocation wraps each call in try/catch, logs errors, and continues.
- **Required plugin protection**: Required plugins (`required: true`) cannot be disabled. The registry endpoint returns a 403 if attempted. This ensures the system always has its core functionality.

### Open questions
1. **Auth policy for the generic plugin dispatch path**: Should the generic `/x/api/v1/plugin/{pluginMachineName}/{callback}` path in the main spec have a single base auth policy (e.g., `authenticated`), with the plugin loader populating per-route policies from each plugin's OpenAPI security declarations into the auth policy map? Or should all plugin routes default to the base policy and plugins enforce stricter auth in their own handlers? I lean toward the former (plugin loader populates the map) so the existing `validateSiteApiRouteAccess` middleware handles enforcement uniformly.
2. **Lifecycle hook points for MVP**: I proposed 14 hook points (onSiteCreated, onSiteDeleted, onSiteCloned, onPageCreated, onPageSaved, onPageDeleted, onManifestSaved, onOutlineSaved, onFileUploaded, onFileDeleted, onUserLogin, onUserLogout, onPluginEnabled, onPluginDisabled). Is this the right scope for MVP, or should we start smaller (e.g., just page/content lifecycle) and expand?
3. **Plugin OpenAPI spec discovery path**: Should plugin OpenAPI specs be served from the system API (`/system/api/v1/plugins/:pluginName/openapi.json`) or the site API (`/x/api/v1/plugin/:pluginName/openapi.json`)? The site API path is more consistent with the generic dispatch pattern, but the system API path keeps plugin management centralized. I lean toward site API for consistency with how plugin routes are accessed.
4. **Which existing features become required core plugins first?** The issue mentions "a small core of plugins that are always loaded and required." Candidates: appDataStorage (from #2474), the app-store brokers, file management, search. Which are truly required (system breaks without them) vs. optional (can be disabled)?
5. **Dependency auto-enable**: When enabling plugin A that depends on B, should the registry auto-enable B (with `?autoEnableDependencies=true`), or require manual enabling in dependency order? Drupal supports both — manual by default, auto with a flag. I propose the same.
6. **Plugin versioning in dependencies**: Should `dependencies` in `plugin.yml` specify version constraints (e.g., `comment-engine: ^1.0.0`), or just machine names? Version constraints add complexity but prevent breakage when a dependency introduces a breaking change. I propose machine-name-only for MVP, version constraints as a future enhancement.

_Plan drafted with Oz via the HAX Issue Research skill._

**Comment by btopro** [comment]
## Implementation plan (UPDATED — supersedes previous comment)

> Drafted via the HAX Issue Research workflow — all open questions resolved with the issue author. This is the finalized ideation-phase architectural plan. Implementation will follow once the approach is ratified.

**Issue:** https://github.com/haxtheweb/issues/issues/1325
**Repo / surface:** `haxcms-nodejs` (primary) + `haxcms-php` (parity) + `HAXiam` (deployment-level config) + `webcomponents` (frontend block registration)

### Problem statement
HAXcms needs a plugin architecture that allows extensions (app-store brokers, commenting engines, xAPI tracking, HAXiam user access, views, etc.) to be autoloaded from directories at multiple config tiers (haxiam core, haxiam config, haxcms core, haxcms config, hax site config), declare their own endpoints via OpenAPI fragments, supply frontend blocks, provide settings forms, enforce Drupal-style dependencies, support pre-action lifecycle hooks that can modify or veto operations, and be enabled/disabled per site via the platform capabilities system. Currently these extensions are hard-coded (e.g., `PROVIDER_DEFINITIONS` in `integrations.js`, `loadBaseAppStore()` in `HAXAppStoreService.js`).

### Current state / context
- **Skeleton system is the closest existing analog**: `haxcms-nodejs/src/lib/skeletonSettings.js:58-73` (`getSkeletonDirectories()`) implements a 3-tier cascade: `user` (`configDirectory/user/skeletons`), `config` (`configDirectory/skeletons`), `core` (`coreConfigPath/skeletons`). `discoverSkeletons()` (line 184) walks these dirs, loads JSON, deduplicates by machine name (first scope wins). `enabledSkeletons.json` in settings tracks enable/disable state. **This is the exact pattern plugins should follow**, extended with executable code and OpenAPI fragments.
- **Config discovery cascade**: `haxcms-nodejs/src/lib/discoverConfigPath.js` walks from cwd to root looking for `_config`/`.haxcmsconfig` dirs with `.isHAXcmsConfig` marker. Home dir fallback. Creates `settings/`, `user/skeletons/`, `skeletons/` dirs on first run (line 38-53).
- **Deployment profiles**: `haxcms-nodejs/src/lib/HAXCMS.js:3096-3105` — `getDeploymentProfile()` returns `single-site`, `self-hosted-multi-site`, or `haxiam-managed`. These control which plugin config tiers are active.
- **Platform features (enable/disable)**: `haxcms-nodejs/src/lib/platformFeatures.js` — `platformAllows(site, capability)` reads from `site.manifest.metadata.platform.features`. Issue #2478 (closed) established the `manifest.metadata.platform` pattern. Plugins register capabilities that flow through this same mechanism.
- **API keys (settings pattern)**: `haxcms-nodejs/src/lib/apiKeys.js` — `SUPPORTED_API_KEY_PROVIDERS` is hard-coded. Reads/writes `settings/apiKeys.json`. This pattern generalizes to plugin-supplied settings.
- **App-store brokers (prime conversion candidate)**: `haxcms-nodejs/src/siteRoutes/v1/integrations.js:19-90` has `PROVIDER_DEFINITIONS` (hard-coded). `haxcms-nodejs/src/lib/HAXAppStoreService.js:10-711` has `loadBaseAppStore()` with embedded JSON strings. **All of this should become plugin-supplied.**
- **Route registration (static)**: `haxcms-nodejs/src/lib/SiteRoutesMap.js:88-108` (`V1_ENTITY_ROUTE_DEFINITIONS`) + `SystemRoutesMap.js` — routes statically registered at startup via `addRouteHandler()`. Plugins dynamically inject routes.
- **OpenAPI spec drives auth**: `haxcms-nodejs/src/app.js:1541-1606` — `readSiteApiAuthPoliciesFromOpenApiSpec()` parses `site-spec.yaml` to determine auth policies per route. Routes not in the spec fail-closed to `authenticated`.
- **Frontend element loading**: `haxcms-nodejs/src/coreConfig/dynamicElementLoader.json` maps tag names to import paths. Plugins contribute entries.
- **PHP parity**: `haxcms-php/system/api.php` uses `SystemApiRouter::dispatch()`. The PHP `_config` directory mirrors the NodeJS config directory.
- **Issue #213 vision**: Plugin distribution as a decentralized "supplier" and "package manager" ecosystem — like apt-get/yum for HAX. Different suppliers for different contexts (EDU, cooking, etc.). CDN alignment with supplier-specific configurations. Start on-prem, work up to registry.

### Proposed approach

**1. Plugin directory cascade (config tiers, first-match-wins by machine name)**

Plugins are housed at the system level across multiple config tiers. A site can override by placing a plugin in its own config, but the primary housing is system-level:
- **HAXiam core**: `{haxiamRoot}/core/plugins` — plugins shipped with HAXiam
- **HAXiam config**: `{haxiamConfig}/plugins` — deployment-level plugins for HAXiam-managed installations
- **HAXcms core**: `{coreConfigPath}/plugins` — plugins shipped with HAXcms (the `system`, `api-system`, `api-site` required plugins live here)
- **HAXcms config**: `{configDirectory}/plugins` — system-wide plugins for self-hosted or HAXiam installations
- **HAX site config**: `{site.siteDirectory}/config/plugins` — per-site plugin overrides

Discovery walks these tiers in order; first match by machine name wins (same dedup pattern as skeletons). This means a site can override a system plugin, or a deployment can override a core plugin.

A small core of always-loaded **required** plugins ensures the system works. Other plugins are optional and can be enabled/disabled **at the site level** (skeletons can pre-enable/disable a batch of plugins, but explicit enable/disable is a site-level operation).

New `haxcms-nodejs/src/lib/pluginLoader.js` (mirroring `skeletonSettings.js`) implements `getPluginDirectories()`, `discoverPlugins()`, `readEnabledPluginMap()`, `writeEnabledPluginMap()`.

**2. Plugin structure (per folder)**

```
plugins/
  comment-engine/
    plugin.yml              # info, capabilities, dependencies, required flag, version
    openapi.yaml            # OpenAPI 3.0 fragment: this plugin's own paths/schemas
    settings.schema.json    # HAXSchema form definition for plugin settings
    node/
      index.js              # exports: routes, hooks, capabilities, roles
      handlers/             # route handler modules
    php/
      CommentEngine.php     # PHP parity handler
    frontend/
      blocks/               # web component JS files
      elements.json         # tag-name -> import-path mappings
      admin/                # optional: admin panel UI for this plugin
```

`plugin.yml` fields:
- `machineName` — unique identifier (sanitized, kebab-case)
- `title`, `description`, `version`, `author` — `version` is informational only (for registry display and future upgrade flows); not enforced in dependency resolution
- `scope` — which config tier this plugin lives in
- `required` — boolean; if true, always loaded and cannot be disabled
- `capabilities` — list of platform capability keys (namespaced, e.g., `comment-engine.hasComments`)
- `dependencies` — list of other plugin machine names (machine-name-only, no version constraints). All plugins have an implicit dependency on the `system` plugin even if not declared.
- `roles` — optional: list of role definitions this plugin provides (e.g., `{ name: 'moderator', capabilities: ['comment-engine.canModerate'] }`)
- `settingsSchema` — path to settings.schema.json
- `openApiSpec` — path to openapi.yaml
- `languages` — list of supported backend languages (e.g., `['node', 'php']`); a plugin can be Node-only, PHP-only, or both

**3. Required core plugins**

Three required plugins form the foundation. They are always loaded and cannot be disabled:
- **`system`** — manages the loading of plugins at the system-wide level. Defines the structure of plugins to be loaded. All other plugins have a built-in implicit dependency on `system` even if not declared in their `dependencies`.
- **`api-system`** — handles loading of OpenAPI specs for system-level transactions (`/system/api/*`). Takes the place of the current static `SystemRoutesMap` OpenAPI handling. Required so the system API itself works.
- **`api-site`** — handles loading of OpenAPI specs for site-level transactions (`/x/api/*`). Takes the place of the current static `SiteRoutesMap` OpenAPI handling. Required so the site API itself works.

These three plugins effectively modularize the current hard-coded route and spec loading, making the core itself plugin-architected.

**4. First optional plugin: `views` (proof of concept for enable/disable)**

`views` is the first optional plugin to test the system's ability to support enabling/disabling functionality. It's a good candidate because it adds to the admin UI, has its own routes, and has frontend pieces. Enabling/disabling `views` exercises the full plugin lifecycle: route registration/unregistration, frontend element loading, capability gating, and admin UI integration.

**5. `debug` plugin (development aid)**

A `debug` plugin that prints console.log statements in NodeJS to the terminal output for all hooks. This lets developers see what hooks are firing when the platform is in use and the transactions involved. This is an optional plugin that can be enabled in development environments. It registers a hook listener on every hook point and logs the hook name, plugin name, and context data.

**6. Generic plugin route dispatch (key architectural decision)**

Plugin routes use a **generic dispatch pattern** rather than merging all plugin OpenAPI specs into one master spec:

- The main `site-spec.yaml` declares a generic path: `/x/api/v1/plugin/{pluginMachineName}/{callback}` with a base security policy.
- When a request matches this pattern, the backend extracts `pluginMachineName` and the sub-path, looks up the plugin's route map (registered by its `node/index.js`), and dispatches to the plugin's handler.
- The plugin's own `openapi.yaml` is **consulted internally** by the backend for request/response schema validation and for per-endpoint auth policy. It is **not merged** into the main spec.
- Each plugin's OpenAPI spec is served at a discovery endpoint: `GET /x/api/v1/plugin/{pluginMachineName}/openapi.json` (and `.yaml`) so clients and LLMs can understand the plugin's specific API surface.

The auth middleware (`validateSiteApiRouteAccess`) is extended so the plugin loader populates the `siteApiAuthPoliciesByMethodAndRoute` map with per-plugin-route entries extracted from each plugin's OpenAPI security declarations. The generic dispatch path has a base policy of `authenticated` (fail-closed), and the plugin loader populates per-route entries (e.g., `authenticated-user` for routes requiring `userTokenHeader`) so the existing fail-closed enforcement works uniformly.

**7. Plugin registry API (system/site split)**

Per the issue author's direction, plugin information is read at the system level, but enable/disable is a site-level operation:

**System-level (read plugin definitions):**
- `GET /system/api/v1/plugins` — list all discovered plugins across all config tiers with scope, required flag, dependencies, version, settings summary
- `GET /system/api/v1/plugins/:pluginName` — plugin definition detail with full info, settings form schema, OpenAPI spec link
- `GET /system/api/v1/plugins/:pluginName/openapi.json` — plugin's OpenAPI spec (JSON)
- `GET /system/api/v1/plugins/:pluginName/openapi.yaml` — plugin's OpenAPI spec (YAML)

**Site-level (enable/disable + settings):**
- `GET /x/api/v1/site/plugins` — list plugins with enabled/disabled status for the current site
- `PATCH /x/api/v1/site/plugins/:pluginName` — enable/disable plugin for this site (Drupal-style dependency enforcement: refuses to disable required plugins; refuses to enable if dependencies are missing; auto-enables dependencies)
- `GET /x/api/v1/site/plugins/:pluginName/settings` — get plugin settings values + HAXSchema form for this site
- `PATCH /x/api/v1/site/plugins/:pluginName/settings` — save plugin settings (validates against schema)

System-level routes are registered in `SystemRoutesMap.js` as admin routes. Site-level routes are registered in `SiteRoutesMap.js` with `authenticated-site` policy.

**8. Dependency enforcement (Drupal-style, auto-enable)**

- A plugin cannot be **enabled** if its declared dependencies are not enabled. The registry endpoint **automatically enables missing dependencies** (since they're listed as dependencies, you can't enable A without B also being enabled).
- When done via frontend, this is a confirmation step: "Hey, this will also be enabled: [list of dependencies]."
- When done via CLI, it respects `--y` / `--no-i` flags to assume the user means to automatically enable the dependencies without prompting.
- A plugin cannot be **disabled** if other enabled plugins depend on it. The registry endpoint returns a 409 with a message listing dependent plugins that must be disabled first.
- Required plugins (`required: true` in `plugin.yml`) cannot be disabled at all.
- Dependencies are **machine-name-only** (no version constraints). `version` is an informational field in `plugin.yml` for registry display and future upgrade flows, but is not enforced during dependency resolution.

**9. Lifecycle hooks (Drupal 7 hook-style, pre-action, can modify AND veto)**

A plugin's `node/index.js` exports a `hooks` object with named hook functions. The `pluginLoader.js` maintains a hook registry and invokes all enabled plugins' hooks for each event. Hooks fire **prior to the operation happening** so plugins can influence it. A hook can **modify the input/output data** AND **optionally veto (block) the operation** by returning a specific error structure.

**Entity CRUD hooks (pre-action, can modify or veto):**

Each entity type (site, page/item, manifest, outline, file, etc.) has Create, Read, Update, Delete hooks:
- `onSiteCreate(site, context)` — before a site is created; hook can modify the site payload or veto
- `onSiteRead(site, context)` — before site data is returned; hook can modify the response
- `onSiteUpdate(site, context)` — before a site is updated; hook can modify the update payload or veto
- `onSiteDelete(siteName, context)` — before a site is deleted; hook can veto or perform cleanup
- `onPageCreate(site, item, context)` — before a page is created
- `onPageRead(site, item, context)` — before page data is returned
- `onPageUpdate(site, item, context)` — before a page is updated
- `onPageDelete(site, itemId, context)` — before a page is deleted
- `onManifestSave(site, context)` — before the manifest is saved
- `onOutlineSave(site, items, context)` — before the outline is saved
- `onFileCreate(site, file, context)` — before a file is uploaded
- `onFileRead(site, file, context)` — before file data is returned
- `onFileDelete(site, filePath, context)` — before a file is deleted

This CRUD pattern mirrors the REST verbs, making it consistent and predictable. As new entities are added (including entities defined as plugins), they follow the same Create/Read/Update/Delete hook pattern.

**Generic REST transaction hooks:**
- `onRESTRequest(request, context)` — before any REST request is processed; hook can modify the request (headers, body, query params) or veto
- `onRESTResponse(response, context)` — before any REST response is sent; hook can modify the response body, status, or headers

These generic hooks let a plugin jump in at the transaction/transform layer, working with the request when it goes up and the response on the way back. This enables cross-cutting concerns (logging, caching, security filtering, transformation) without needing entity-specific hooks.

**Veto mechanism:** A hook returns `{ veto: true, status: 403, message: 'Operation blocked by plugin X' }` to block the operation. The backend returns the specified status/message to the client. If no veto, the hook can return `{ modifiedData: ... }` to replace the input/output data, or return nothing/`{}` to let the operation proceed unmodified.

Hooks are invoked in plugin load order. Multiple plugins' hooks for the same event are called sequentially; each can modify the data and the next receives the modified version (chaining). Errors in hooks are logged but don't crash the operation (unless a hook explicitly vetos).

**10. Platform capabilities integration**

Plugins declare `capabilities` in `plugin.yml`. When enabled for a site, these flow into `site.manifest.metadata.platform` (or a new `site.manifest.metadata.plugins` namespace). The existing `platformAllows(site, capability)` function is extended to check plugin-provided capabilities. Example: `comment-engine` provides `hasComments`; frontend checks `store.platformAllows('hasComments')` before rendering.

**11. Plugin-defined roles**

Plugins can declare `roles` in `plugin.yml`. The core system defines base roles (site admin, site editor, site viewer). Plugins can define additional roles (e.g., `comment-engine` could define a `moderator` role with `comment-engine.canModerate` capability). These roles are integrated into the site user permissions system (connecting to issue #2479 — user permissions in HAXiam context). A site admin has access to all plugin capabilities by default.

**12. Settings management (plugin-supplied forms + admin UI)**

Plugin `settings.schema.json` is a HAXSchema form definition. Plugin settings endpoints serve this as a form. Settings stored in `settings/pluginSettings/{pluginName}.json` (system-level) or `{site.siteDirectory}/config/pluginSettings/{pluginName}.json` (site-level). The `apiKeys.js` pattern is generalized — instead of a hard-coded `SUPPORTED_API_KEY_PROVIDERS`, a provider plugin's settings schema declares its `apiKey` field.

**Plugin admin UI** follows the path cascade: `admin -> plugins -> {pluginName}`. This UI lets site admins:
- Enable and disable plugins (with dependency auto-enable confirmation)
- Click to edit the specific settings of each plugin (renders the HAXSchema form from `settings.schema.json`)
- View plugin details (version, author, description, capabilities, dependencies)
- Access plugin-specific admin panels (e.g., comment moderation for `comment-engine`)

This pattern builds out the admin UI for plugins broadly — the same `admin -> plugins -> {pluginName}` path works for any plugin's settings and admin functionality.

**13. OpenAPI fragment handling (per plugin, not merged)**

Plugin OpenAPI specs are **not merged** into the main spec. Instead:
- The main `site-spec.yaml` has the generic `/x/api/v1/plugin/{pluginMachineName}/{callback}` path with base `authenticated` security.
- Each plugin's `openapi.yaml` is self-contained with its own paths, schemas, and security schemes.
- At startup, the plugin loader reads each plugin's OpenAPI security declarations and populates the auth policy map so `validateSiteApiRouteAccess` enforces the correct policy per plugin route.
- Plugin OpenAPI specs are served at `/x/api/v1/plugin/{pluginMachineName}/openapi.json` and `.yaml` for discovery.
- The existing `/x/api/openapi` endpoint continues to serve the main spec (which now includes the generic plugin path pattern and a `plugins` directory listing linking to each plugin's spec).

**14. Plugin execution (main process, trusted)**

Plugin NodeJS modules run in the main process (no VM sandboxing). Plugins are trusted as part of the community/platform ecosystem — installed from a future registry or manually downloaded by developers. Core plugins shipped with the platform leverage the same architecture. This matches how `require()` already works for core modules. Provenance verification (checksums, signed manifests) is a future enhancement.

**15. Plugin distribution (on-prem first, registry later per #213)**

Per the issue author's direction and issue #213, plugin distribution starts with on-prem loading (manual copy / git clone / npm install into the config tier directories) and works up to a registry/supplier ecosystem. The directory structure supports all distribution methods. Issue #213's vision of decentralized "suppliers" and "package managers" (like apt-get/yum) is the long-term goal. The MVP focuses on the loading architecture; the registry/supplier UI is future work.

**16. Conversion candidates (proof of concept)**

Per the issue comment and the author's direction:
- **Required core plugins (always loaded):** `system`, `api-system`, `api-site` — modularize the current hard-coded route/spec loading
- **First optional plugin (enable/disable test):** `views` — has routes, frontend, admin UI; exercises the full plugin lifecycle
- **Development aid:** `debug` — logs all hooks to console
- **Future conversion candidates:** `app-store-brokers` (YouTube/Vimeo/etc.), `comment-engine` (bridge to #2474), `xapi-tracker`, `haxiam-access` (PHP-only initially)

### Files likely affected (new)
- `haxcms-nodejs/src/lib/pluginLoader.js` — **NEW** — plugin discovery, loading, cascade logic, dependency enforcement, hook registry
- `haxcms-nodejs/src/lib/pluginSettings.js` — **NEW** — plugin settings storage (generalizes `apiKeys.js`)
- `haxcms-nodejs/src/lib/pluginRouteDispatcher.js` — **NEW** — generic `/x/api/v1/plugin/:pluginMachineName/*` route handler that dispatches to plugin handlers
- `haxcms-nodejs/src/systemRoutes/v1/plugins.js` — **NEW** — system-level plugin registry (list, detail, OpenAPI)
- `haxcms-nodejs/src/siteRoutes/v1/plugins.js` — **NEW** — site-level plugin enable/disable + settings
- `haxcms-nodejs/src/coreConfig/plugins/system/` — **NEW** — required `system` plugin
- `haxcms-nodejs/src/coreConfig/plugins/api-system/` — **NEW** — required `api-system` plugin
- `haxcms-nodejs/src/coreConfig/plugins/api-site/` — **NEW** — required `api-site` plugin
- `haxcms-nodejs/src/coreConfig/plugins/views/` — **NEW** — first optional plugin (converted from current hard-coded views)
- `haxcms-nodejs/src/coreConfig/plugins/debug/` — **NEW** — debug plugin (logs hooks to console)
- `webcomponents/elements/` — admin UI components for `admin -> plugins -> {pluginName}` panel

### Files likely affected (modified)
- `haxcms-nodejs/src/app.js` — call plugin discovery + route registration + hook registration at startup (~line 1181 area); register generic plugin dispatch route; add hook invocation points at all entity CRUD operations
- `haxcms-nodejs/src/lib/SiteRoutesMap.js` — support late route registration from plugins; add generic plugin dispatch path + site-level plugin management routes
- `haxcms-nodejs/src/lib/SystemRoutesMap.js` — add system-level plugin registry routes
- `haxcms-nodejs/src/lib/HAXCMS.js` — ensure `config/plugins` and `settings/pluginSettings` dirs exist (mirrors skeleton dir creation at line 2816-2828); add hook invocation points at all lifecycle events (site create/read/update/delete, page CRUD, file CRUD, manifest/outline save, REST request/response)
- `haxcms-nodejs/src/lib/platformFeatures.js` — extend `platformAllows()` to check plugin capabilities; integrate plugin-defined roles
- `haxcms-nodejs/src/lib/apiKeys.js` — generalize to read supported providers from discovered plugins
- `haxcms-nodejs/src/siteRoutes/v1/integrations.js` — refactor to load provider definitions from `app-store-brokers` plugin (future)
- `haxcms-nodejs/src/lib/HAXAppStoreService.js` — refactor `loadBaseAppStore()` to load from plugin data (future)
- `haxcms-nodejs/src/openapi/site-spec.yaml` — add generic `/x/api/v1/plugin/{pluginMachineName}/{callback}` path with base security
- `haxcms-nodejs/src/siteRoutes/v1/views.js` — extract into `views` plugin (first optional plugin conversion)

### Considerations / risks
- **DDD design system**: N/A for backend. Frontend blocks shipped by plugins should use DDD tokens. The admin UI (`admin -> plugins -> {pluginName}`) should use DDD.
- **Accessibility/WCAG**: Plugin-supplied frontend blocks and admin UI must follow the same accessibility audits as core elements.
- **HAXSchema/HAX capability**: Plugin settings forms use HAXSchema. Plugin-supplied blocks must have `haxProperties` to be authorable in HAX.
- **New elements via CLI**: Plugin frontend blocks that are new web components must be created via `hax webcomponent` CLI, never by hand.
- **Backend parity (PHP)**: The issue requires "nodejs and php based integration." Each plugin has a `php/` directory with a PHP handler. Plugins can be Node-only, PHP-only, or both — the shared spec (plugin.yml + openapi.yaml) provides a unified contract.
- **NO top-level monorepo build** (standing rule).
- **NO optional chaining (`?.`)** — the Polymer parser breaks on it (standing rule).
- **`globalThis` over `window`** for global-scope JS (standing rule).
- **Security — plugin code trusted in main process**: Plugins run in the main Node process with full access. For now, plugins are trusted as community/platform code — same trust model as npm packages. Provenance verification is a future enhancement.
- **Hook veto power**: Hooks can veto operations. A misbehaving plugin could block critical operations (e.g., a buggy `onSiteDelete` hook preventing site deletion). The `debug` plugin helps diagnose this. Consider an "emergency disable all hooks" escape hatch for admins.
- **Hook chaining**: Multiple plugins' hooks for the same event are called sequentially, each receiving the previous plugin's modifications. Plugin load order matters. Required plugins load first, then optional plugins in config-tier order.
- **Route name collision**: Since routing is namespaced by `pluginMachineName` (`/x/api/v1/plugin/{pluginMachineName}/{callback}`), collisions between plugins are impossible. Within a single plugin, the route map must not have duplicate keys — the loader validates this.
- **Capability namespace collision**: Plugin capability keys are namespaced by plugin machine name (e.g., `comment-engine.hasComments`).
- **Reserved `x/` routes**: Plugin routes use `/x/api/v1/plugin/{pluginMachineName}/...` — no conflict with reserved `x/` internal paths.
- **site.json `metadata.site.name` alignment**: Site-scoped plugins live under `{site.siteDirectory}/config/plugins` which is aligned with the site name.
- **Performance**: Plugin discovery at startup is cached. The discovered plugin list, parsed OpenAPI specs, and hook registry are cached in memory. Invalidation on admin enable/disable or file watcher event.
- **Migration path**: The `system`, `api-system`, and `api-site` required plugins are created first (extracting current hard-coded logic). Then `views` is extracted as the first optional plugin to test enable/disable. No breaking changes to existing API consumers during transition.
- **Required plugin protection**: Required plugins (`required: true`) cannot be disabled. The registry endpoint returns a 403 if attempted.
- **`views` as first optional plugin**: Extracting `views` from hard-coded routes into a plugin is the proof-of-concept for the enable/disable lifecycle. If `views` can be disabled and the system still works (just without views functionality), the plugin architecture is proven.

### Open questions
1. **Hook ordering and priority**: When multiple plugins register hooks for the same event, what determines the invocation order? Proposed: required plugins first (system, api-system, api-site), then optional plugins in config-tier order (haxiam core -> haxiam config -> haxcms core -> haxcms config -> site config), then alphabetical by machine name within each tier. Should plugins be able to declare a `priority` field to override this order?
2. **Emergency hook disable**: Should there be an admin escape hatch to disable all hooks (e.g., an env var `HAXCMS_DISABLE_HOOKS=1` or a system setting) in case a misbehaving plugin's hook veto blocks critical operations? I propose yes, for safety.
3. **Plugin-defined roles and HAXiam integration**: The plan says plugins can define roles (e.g., `moderator`). How do these integrate with HAXiam's user management (issue #2479)? Is there a role sync mechanism, or does HAXiam read plugin-defined roles at startup? This is a bridge point between #1325 and #2479.
4. **`api-system` and `api-site` plugin boundaries**: The `api-system` plugin handles `/system/api/*` OpenAPI spec loading and the `api-site` plugin handles `/x/api/*`. Should the generic plugin dispatch route (`/x/api/v1/plugin/...`) be part of `api-site`'s responsibility, or a separate concern owned by the `system` plugin? I lean toward `api-site` owning it since it's a site API path.

_Plan drafted with Oz via the HAX Issue Research skill._

**Comment by btopro** [comment]
## Plan audit feedback (Oz)

> Audit of the updated implementation plan above. File/line references in the plan were spot-checked against `haxcms-nodejs` main and are accurate (`skeletonSettings.js`, `platformFeatures.js`, `getDeploymentProfile()` HAXCMS.js:3096, `readSiteApiAuthPoliciesFromOpenApiSpec()` app.js:1541, etc.). The plan is well-grounded and implementable. Below: strengths, risks, and proposed resolutions for future context.

### Strengths
- Anchoring on the skeleton cascade reuses a proven core pattern instead of a new abstraction.
- Generic dispatch (`/x/api/v1/plugin/{name}/{callback}`) instead of merging OpenAPI specs keeps the main spec stable, eliminates route collisions by construction, and preserves fail-closed auth.
- `views` as first optional plugin gives a falsifiable success criterion: disable it and the system still works.
- `debug` plugin doubles as validation of the hook registry.

### Risks identified + proposed resolutions

**1. Pre-action hook model is the riskiest part (and understated).**
- Read hooks (`onSiteRead`/`onPageRead`/`onFileRead`) sit on the hottest paths; one slow hook degrades every page load.
- *Resolution:* plugins must declare which hooks they subscribe to in `plugin.yml`; the loader builds a per-hook subscriber index so unsubscribed hook points (especially reads) cost zero at runtime.
- Error semantics contradiction: "errors are logged but don't crash the operation" is dangerous for hooks that *modify* data mid-chain.
- *Resolution:* an error in a modifying hook discards that plugin's modification and continues the chain with the prior data; add a per-hook timeout so a hung hook can't wedge requests.
- `onRESTRequest`/`onRESTResponse` overlap with Express middleware.
- *Resolution:* implement them as thin wrappers over Express middleware; explicitly skip `onRESTResponse` for streamed/file responses.

**2. Bootstrap chicken-and-egg for `system`/`api-system`/`api-site`.** Something must load the `system` plugin itself, and if `api-site` owns site-spec loading, a malformed `api-site` yaml would take down the site API.
- *Resolution:* `pluginLoader.js` is irreducible core code. The three required plugins are real extractions of current logic but are loaded from hardcoded paths with fail-fast validation (server refuses to start if broken) rather than discovered through the cascade. Document this bootstrap order in the plan so they don't become label-only wrappers.

**3. Site-tier override of executable code is a privilege-escalation path.** "First match wins" is fine for JSON skeletons; plugins contain Node code, and site directories may be user-writable in `haxiam-managed` deployments.
- *Resolution:* site tier can only override config/settings/enable-state — never executable code. Code loading from site config directories is disabled entirely under the `haxiam-managed` deployment profile.

**4. Per-site enable/disable vs. process-global hooks/routes.** "Invokes all enabled plugins' hooks" doesn't say enabled *for which site*, and some hooks (e.g. `onSiteCreate`) have no site context yet.
- *Resolution:* hook registry is global; invocation filters against a per-site enabled map cached on the site object, invalidated on enable/disable. System-level hooks (no site context) filter against system-tier enablement.

**5. PHP parity is asserted, not planned.** Every Node feature (hook chaining, veto, dispatch, auth policy population) needs a PHP twin — roughly double the work.
- *Resolution:* Node-first MVP; PHP parity is explicitly deferred until the architecture is ratified and proven with `views`. `plugin.yml` + `openapi.yaml` remain the language-neutral contract so PHP can follow without schema changes.

**6. Settings secrets can leak via site directories.** Site-level `pluginSettings/*.json` may be published/synced (git, static publish).
- *Resolution:* settings schema fields marked `secret: true` are stored at system level only (per current `apiKeys.js` pattern) and masked in GET responses.

### Answers to the plan's open questions
1. **Hook ordering:** proposed order (required → config-tier order → alphabetical) is fine; add an integer `priority` field (default 0) now — cheap, and Drupal history says it'll be needed immediately.
2. **Emergency hook disable:** yes — `HAXCMS_DISABLE_HOOKS=1`, but scoped to *optional* plugin hooks only, so the escape-hatch state is still well-defined.
3. **Plugin-defined roles / HAXiam:** defer role sync out of MVP; ship `roles` in `plugin.yml` as declared-but-inert metadata until #2479 lands, so the schema is stable now.
4. **Dispatch ownership:** `api-site` owns `/x/api/v1/plugin/...` — contingent on the bootstrap resolution in point 2 above.

### Smaller nits
- Drop `scope` from `plugin.yml` — the tier is determined by which directory the plugin was found in; a self-declared scope that disagrees with location is a validation headache.
- Use 409 for both dependency-conflict and required-plugin-disable refusals (neither is an authorization failure; 403 is misleading).
- Skip file watchers for MVP — cache-bust on admin enable/disable actions only; watchers across N site dirs are expensive.
- Purge the v1 past-tense hook names (`onPageSaved` etc.) from whatever doc gets ratified so implementers don't accidentally build post-hooks.

**Net:** resolve the hook execution model details (read-path cost, error semantics, per-site filtering) and the bootstrap definition of the three required plugins before ratifying; everything else is polish.

_Audit performed with Oz._
Co-Authored-By: Oz <oz-agent@warp.dev>

**Comment by btopro** [comment]
@Hellothereobi123

**Comment by btopro** [comment]
https://mglaman.dev/blog/simplytestme-can-launch-drupal-cms-site-templates example to read and see schema
